src/festim/stepsize.py

```diff
@@ -67,16 +91,55 @@ def cutback_factor(self, value):
 
         self._cutback_factor = value
 
+    @property
+    def max_stepsize(self):
+        return self._max_stepsize
+    
+    @max_stepsize.setter
+    def max_stepsize(self, value):
+        if value is not None and not callable(value):
+            if value < self.initial_value:
+                raise ValueError("maximum stepsize cannot be less than initial stepsize")
+            
+        self._max_stepsize = value
+
+    def get_max_stepsize(self, t):
+        """
+        Returns the maximum stepsize at time t. 
+        
+        Args:
+            t (float): the current time
+        
+        Returns:
+            float or None: the maximum stepsize at time t
+        """
+        if callable(self._max_stepsize):
+            return self._max_stepsize(t)
+        return self._max_stepsize
+
     def modify_value(self, value, nb_iterations, t=None):
         if not self.is_adapt(t):
             return value
 
+        max_step = self.get_max_stepsize(t)
+
         if nb_iterations < self.target_nb_iterations:
-            return value * self.growth_factor
+            updated_value = min(value * self.growth_factor, max_step if max_step is not None else value * self.growth_factor)
         elif nb_iterations > self.target_nb_iterations:
-            return value * self.cutback_factor
+            updated_value = value * self.cutback_factor
         else:
-            return value
+            updated_value = value
```

Comment:
This will only cap the stepsize if `nb_iterations` is less than the target. But imagine a case where at each iteration you always are equal or above. You would could end up with updated_value > max stespize.

I would do:

```python
        if nb_iterations < self.target_nb_iterations:
            updated_value = value * self.growth_factor
        elif nb_iterations > self.target_nb_iterations:
            updated_value = value * self.cutback_factor
        else:
            updated_value = value

        max_step = self.get_max_stepsize(t)
        if max_step:
            if updated_value > max_step:
                updated_value = max_step

Reply:
To check this simply run this test:

```python
    my_stepsize = F.Stepsize(initial_value=1)
    my_stepsize.max_stepsize = 1
    my_stepsize.growth_factor = 1.1
    my_stepsize.target_nb_iterations = 4

    current_value = 2
    new_value = my_stepsize.modify_value(
        value=current_value,
        nb_iterations=my_stepsize.target_nb_iterations,
    )

    expected_value = 1
    assert new_value == expected_value
```

It returns an assertion error:

```
>       assert new_value == expected_value
E       assert 2 == 1

test/test_stepsize.py:98: AssertionError
```